src/hugr/views/sibling.rs

```diff
@@ -0,0 +1,572 @@
+//! Views for HUGR sibling subgraphs.
+//!
+//! Views into subgraphs of HUGRs within a single level of the
+//! hierarchy, i.e. within a sibling graph. Such a subgraph is
+//! represented by a parent node, of which all nodes in the sibling subgraph are
+//! children, as well as a set of edges forming the subgraph boundary.
+//! The boundary must be fully contained within the sibling graph of the parent.
+//!
+//! Sibling subgraphs complement [`super::HierarchyView`]s in the sense that the
+//! latter provide views for subgraphs defined by hierarchical relationships,
+//! while the former provide views for subgraphs within a single level of the
+//! hierarchy.
+//!
+
+// TODO:
+// //! This module exposes the [`SiblingView`] trait, which is currently
+// //! implemented by the [`SiblingSubgraph`] struct in this module, as well as
+// //! the [`SiblingGraph`] hierarchical view.
+
+use itertools::{Either, Itertools};
+use portgraph::{algorithms::ConvexChecker, view::Subgraph, PortView};
+use thiserror::Error;
+
+use crate::{
+    ops::{OpTag, OpTrait},
+    types::{AbstractSignature, EdgeKind},
+    Direction, Hugr, Node, Port, SimpleReplacement,
+};
+
+use super::{sealed::HugrInternals, HugrView};
+
+/// A non-empty convex subgraph of a HUGR sibling graph.
+///
+/// A HUGR region in which all nodes share the same parent. Unlike
+/// [`super::SiblingGraph`],  not all nodes of the sibling graph must be
+/// included. A convex subgraph is always an induced subgraph, i.e. it is defined
+/// by a set of nodes and all edges between them.
+
+/// The incoming boundary (resp. outgoing boundary) is given by the input (resp.
+/// output) ports of the subgraph that are linked to nodes outside of the subgraph.
+/// The signature of the sugbraph is then given by the types of the incoming
+/// and outgoing boundary edges. Given a replacement with the same signature,
+/// a [`SimpleReplacement`] can be constructed to rewrite the subgraph with the
+/// replacement.
+///
+/// The ordering of the nodes in the subgraph is irrelevant to define the convex
+/// subgraph, but it determines the ordering of the boundary signature.
+#[derive(Clone, Debug)]
+pub struct SiblingSubgraph<'g, Base> {
+    /// The underlying Hugr.
+    base: &'g Base,
+    /// The nodes of the induced subgraph.
+    nodes: Vec<Node>,
+}
+
+impl<'g, Base: HugrInternals> SiblingSubgraph<'g, Base> {
+    /// A sibling subgraph from a HUGR.
+    ///
+    /// The subgraph is given by the sibling graph of the root. If you wish to
+    /// create a subgraph from another root, wrap the argument `region` in a
+    /// [`super::SiblingGraph`].
+    ///
+    /// This panics if the sibling graph is empty.
+    pub fn from_sibling_graph(sibling_graph: &'g Base) -> Self
+    where
+        Base: HugrView,
+    {
+        let root = sibling_graph.root();
+        let nodes = sibling_graph.children(root).collect_vec();
+        if nodes.is_empty() {
+            panic!("Empty graph");
+        }
```

Comment:
I'd rather return an `InvalidSubgraph` for an user's error. (Same for `from_dataflow_graph`)

Reply:
That would mean an empty sibling subgraph, i.e. without even Input/Output nodes right? Isn't that an invalid Hugr? In which case panicking is acceptable??

Reply:
AFAICS this method is *identical* to `from_dataflow_graph` so we can probably remove it?

Reply:
I've gone with @aborgna-q's suggestion. The difference between `from_sibling_graph` and `from_dataflow_graph` is that the latter takes the input/output nodes as inputs and outputs, whereas the former takes the whole graph as-is. As such, an empty graph isn't invalid.